go-controller/observability-lib/sampledecoder/sample_decoder.go

```diff
@@ -0,0 +1,290 @@
+		case libovsdbops.NetpolNodeOwnerType:
+			msg = "default allow from local node policy, direction ingress"
+		case libovsdbops.NetworkPolicyOwnerType:
+			msg = fmt.Sprintf("network policy %s, direction %s", o.ExternalIDs[libovsdbops.ObjectNameKey.String()], o.ExternalIDs[libovsdbops.PolicyDirectionKey.String()])
+		case libovsdbops.NetpolNamespaceOwnerType:
+			msg = fmt.Sprintf("network policies isolation in namespace %s, direction %s", o.ExternalIDs[libovsdbops.ObjectNameKey.String()], o.ExternalIDs[libovsdbops.PolicyDirectionKey.String()])
+		case libovsdbops.EgressFirewallOwnerType:
+			msg = fmt.Sprintf("egress firewall in namespace %s", o.ExternalIDs[libovsdbops.ObjectNameKey.String()])
+		case libovsdbops.UDNIsolationOwnerType:
+			msg = fmt.Sprintf("UDN isolation of type %s", o.ExternalIDs[libovsdbops.ObjectNameKey.String()])
+		}
+		return fmt.Sprintf("%s by %s", action, msg)
+	default:
+		return ""
+	}
+}
+
+func (d *SampleDecoder) DecodeCookieBytes(cookie []byte) (string, error) {
+	if uint64(len(cookie)) != CookieSize {
+		return "", fmt.Errorf("invalid cookie size: %d", len(cookie))
+	}
+	c := Cookie{}
+	err := binary.Read(bytes.NewReader(cookie), SampleEndian, &c)
+	if err != nil {
+		return "", err
+	}
+	return d.DecodeCookieIDs(c.ObsDomainID, c.ObsPointID)
+}
+
+func (d *SampleDecoder) DecodeCookie8Bytes(cookie [8]byte) (string, error) {
+	c := Cookie{}
+	err := binary.Read(bytes.NewReader(cookie[:]), SampleEndian, &c)
+	if err != nil {
+		return "", err
+	}
+	return d.DecodeCookieIDs(c.ObsDomainID, c.ObsPointID)
+}
+
+func getGroupID(groupID *int) string {
+	if groupID == nil {
+		return "unset"
+	}
+	return fmt.Sprintf("%d", *groupID)
+}
+
+func (d *SampleDecoder) AddCollector(collectorID, groupID int, ownerName string) error {
+	if d.ovsdbClient == nil {
+		return fmt.Errorf("OVSDB client is not initialized")
+	}
+	// find existing collector with the same ID
+	collectors := []*ovsdb.FlowSampleCollectorSet{}
+	err := d.ovsdbClient.WhereCache(func(item *ovsdb.FlowSampleCollectorSet) bool {
+		return item.ID == collectorID
+	}).List(context.Background(), &collectors)
+	if err != nil {
+		return fmt.Errorf("failed finding existing collector: %w", err)
+	}
+	if len(collectors) > 0 && (collectors[0].ExternalIDs["owner"] != ownerName ||
+		collectors[0].LocalGroupID == nil || *collectors[0].LocalGroupID != groupID) {
+		return fmt.Errorf("requested collector with id=%v already exists "+
+			"with the external_ids=%+v, local_group_id=%v", collectorID, collectors[0].ExternalIDs["owner"], getGroupID(collectors[0].LocalGroupID))
+	}
+
+	// find br-int UUID to attach collector
+	bridges := []*ovsdb.Bridge{}
+	err = d.ovsdbClient.WhereCache(func(item *ovsdb.Bridge) bool {
+		return item.Name == bridgeName
+	}).List(context.Background(), &bridges)
+	if err != nil || len(bridges) != 1 {
+		return fmt.Errorf("failed finding br-int: %w", err)
+	}
+
+	ops, err := d.ovsdbClient.Create(&ovsdb.FlowSampleCollectorSet{
```

Comment:
What happens if the collector already existed (with the right configuration)?  Won't this Create fail with index uniqueness violation because there's already an entry with the same ID and bridge UUID:

https://github.com/openvswitch/ovs/blob/main/vswitchd/vswitch.ovsschema#L577

Reply:
there is a part before starting with `// find existing collector with the same ID` that is supposed to find any valid existing collectors. It searches by collectorID, so when we come to creating a new collector we know same ID doesn't exist.
But if e.g. someone managed to insert collector with the same ids in the meantime, it will fail and it is the desired outcome.

Reply:
Ah, ok, I totally missed that.